python/objToJSON.c

```diff
@@ -579,7 +592,7 @@ void Object_beginTypeContext (JSOBJ _obj, JSONTypeContext *tc)
     return;
   }
   else
-  if (PyString_Check(obj))
+  if (PyString_Check(obj) && !PyObject_HasAttrString(obj, "__json__"))
```

Comment:
So the idea is a bit that you can have a magical string, which behaves in one way when used as a string, but another (optimized) when used as a JSON. Specifically, you can have a JSON serialization, which is a string, but also when serialized, it is not converted to a string (so no double encoding), but output directly, because it already contains a proper JSON serialization.

Example use for me is how to properly serialize GeoJSON Django fields. I [create a GeoJSON string class](https://github.com/wlanslovenija/nodewatcher/blob/6e5995cde6573b5650a4a274ead2a533e42f457a/nodewatcher/core/api/fields.py#L131), and then [returned it as an optimized value](https://github.com/wlanslovenija/nodewatcher/blob/6e5995cde6573b5650a4a274ead2a533e42f457a/nodewatcher/core/api/fields.py#L164). In this way existing codebase which expects strings work well, but when the time for serialization to JSON  comes, one can quickly render JSON, because it is already there. Moreover, if you serialize to XML, instead, then the string is used as string, so you get embedded JSON inside XML, which is also good in the case of GeoJSON.